### DISCUSSION TITLE
About device_type_adapter

### DISCUSSION THREAD

**Discussion by Lafatalite** [pyri-project/pyri-core#16] [Q&A]
Hello, I have a question about adapters. I want to transfer data from a robot on Computer Two to the front end using Pyri, and I'm modifying the adapter, but it's not working. How do I write it? I noticed that your Robotic adapter reads wire-type data directly and parses some flags before wrapping it in the PyriDeviceTypeAdapterExtendedState type. What should I pay attention to when passing the p_value?
![image](https://github.com/user-attachments/assets/3c09dda8-d3d0-4ef6-8a8b-10acf14169a0)
This is my own service and corresponding architecture. On Computer 2, my RobDef defines a JointState type that is passed via wire in a JointState type, which I also defined myself.
![image](https://github.com/user-attachments/assets/158d90bb-efbe-4bbf-b4d6-37968b003e45)
![image](https://github.com/user-attachments/assets/53327633-7bfa-4ad6-8168-89ba0582bf48)
This is my adapter's code. It currently doesn't have any effect, and when I print the content of `self.core.devices_states.devices_states[current_robot]`, I find that the state is None.
![image](https://github.com/user-attachments/assets/d3981a72-2065-4c3e-9417-81c2bbe8dcd7)

**Comment by johnwason** [comment]
The device type adapter is used with the devices-states service. It is intended to aggregate the state data coming from multiple devices and pass it to the user interface, so the user interface doesn't get bogged down trying to communicate with lots of devices. 

https://github.com/pyri-project/pyri-devices-states

The user interface constantly updates the aggregated state. You can see an example of it being used on the client side:

https://github.com/pyri-project/pyri-robotics-browser/blob/549fecff9b8b291be2a2f499a9f22eecf5fdc069/src/pyri/robotics_browser/components/jog_component.py#L641

**Reply by Lafatalite** [reply]
I want to load a custom .robdef file. I created an RR program on computer2 and loaded this .robdef file. However, when I tried to resolve the file name in the adapter file in PyRI, the frontend displayed the error "Could not resolve named type xxx.xxx.xxx.xxxx." Through debugging, I found that the information from computer2 had already been passed on, so the issue might be occurring during the VarValue resolution. How can I resolve this problem? Currently, I'm able to pass the information using predefined types from the robotraconteur_standard_robdef file, and that works fine.
![image](https://github.com/user-attachments/assets/b14b9e65-0d4d-4938-99f6-f344276ceed1)
![image](https://github.com/user-attachments/assets/44d809c1-e22a-4f78-bd52-8b7d286426dd)

**Reply by johnwason** [reply]
To use a custom type with devices-states you need to have a plugin that provides the robdef to the rest of the system. An example is here:

https://github.com/pyri-project/pyri-variable-storage/blob/master/src/pyri/variable_storage/robdef.py

https://github.com/pyri-project/pyri-variable-storage/blob/1edafae7686bda8f69f650ae1524b9b3ef3fe653/pyproject.toml#L39
